dwave/preprocessing/include/dwave/presolve.h

```diff
@@ -145,15 +56,134 @@ class Presolver {
     /// Return a const reference to the held constrained quadratic model.
     const model_type& model() const;
 
-    /// Return a const reference to the held postsolver.
-    const Postsolver<bias_type, index_type, assignment_type>& postsolver() const;
+    /// Return a sample of the original CQM from a sample of the reduced CQM.
+    template<class T>
+    std::vector<T> restore(std::vector<T> reduced) const {
+        return model_.restore(std::move(reduced));
+    }
 
  private:
     static constexpr double FEASIBILITY_TOLERANCE = 1.0e-6;
     static constexpr double INF = 1.0e30;
 
-    model_type model_;
-    Postsolver<bias_type, index_type, assignment_type> postsolver_;
+    // We want to control access to the model in order to track changes,
+    // so we create a rump version of the model.
+    class ModelView : private model_type {
+     public:
+        ModelView() = default;
+        ~ModelView() = default;
+
+        explicit ModelView(model_type&& model) : model_type(model) {}
+
+        // Const methods are all safe to expose
+        using model_type::num_constraints;
+        using model_type::num_variables;
+        using model_type::lower_bound;
+        using model_type::upper_bound;
+        using model_type::vartype;
+
+        // We can expose the expression access methods because changes to
+        // the individual expressions don't get tracked
+        using model_type::constraint_ref;
+        using model_type::constraints;
+
+        // Adding/removing constraints isn't tracked
+        using model_type::add_linear_constraint;
+        using model_type::remove_constraint;
+
+        // The bound changes don't get tracked
+        using model_type::set_lower_bound;
+        using model_type::set_upper_bound;
+
+        // Expose the objective. Changes don't get tracked
+        auto& objective() { return static_cast<model_type*>(this)->objective; }
+        const auto& objective() const { return static_cast<model_type*>(this)->objective; }
+
+        // Expose the model as a const reference
+        const model_type& model() const { return *this; }
+
+        // Clear the model, but not the transforms queue.
+        model_type detach_model() {
```

Comment:
Here a question too: how are we sure that swapping `*this` with `cqm` will keep the `transforms_` vector?
I'm thinking about when calling this
```ModelView.detach_model()```
not sure if I'm clear..

Reply:
Because we've only implemented swap, but agree we could make it more explicit and head off some future bugs by doing
```c++
swap(static_cast<model_type&>(*this), cqm)
```

Reply:
and here, why not
```
swap(*this, static_cast<model_type&>((cqm))
```
? 
Unless I'm, not understanding things..

Reply:
oh wait, that will actually delete the transforms_ vector

Reply:
```c++
swap(*this, static_cast<model_type&>((cqm))
```
is just the same as what we're doing now. It will use https://github.com/dwavesystems/dimod/blob/c5ee164c61995f99b01dac77e9a6b3e4c06d03bb/dimod/include/dimod/constrained_quadratic_model.h#L214 which uses https://github.com/dwavesystems/dimod/blob/c5ee164c61995f99b01dac77e9a6b3e4c06d03bb/dimod/include/dimod/constrained_quadratic_model.h#L223-L242 so it won't touch `transforms_`.

Reply:
I don't think it's confusing, it's just that I'm using this as a learning experience lol